src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/MonoDroidMarkStep.cs

```diff
@@ -12,17 +15,203 @@ namespace MonoDroid.Tuner
+
+		public override void Process (LinkContext context)
+		{
+			marshalTypes.Clear ();
+			base.Process (context);
+
+			if (UpdateMarshalTypes ())
+				base.Process (context);
+		}
+
+		bool UpdateMarshalTypes ()
+		{
+			MethodDefinition registerMethod;
+			HashSet<string> markedMethods = new HashSet<string> ();
+			var updated = false;
+
+			foreach (var type in marshalTypes) {
+				registerMethod = null;
+				markedMethods.Clear ();
+
+				foreach (var method in type.Methods) {
+					if (method.Name == "__RegisterNativeMembers") {
+						registerMethod = method;
+
+						continue;
+					}
+
+					if (_context.Annotations.IsMarked (method))
+						markedMethods.Add (method.Name);
+				}
+
+				if (registerMethod == null || markedMethods.Count <= 0)
+					continue;
+
+				updated |= UpdateMarshalRegisterMethod (registerMethod, markedMethods);
+			}
+
+			return updated;
+		}
+
+		static bool IsLdcI4 (Instruction instruction, out int intValue)
+		{
+			intValue = 0;
+
+			if (instruction.OpCode == OpCodes.Ldc_I4_0)
+				intValue = 0;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_1)
+				intValue = 1;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_2)
+				intValue = 2;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_3)
+				intValue = 3;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_4)
+				intValue = 4;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_5)
+				intValue = 5;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_6)
+				intValue = 6;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_7)
+				intValue = 7;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_8)
+				intValue = 8;
+			else if (instruction.OpCode == OpCodes.Ldc_I4)
+				intValue = (int) instruction.Operand;
+			else if (instruction.OpCode == OpCodes.Ldc_I4_S)
+				intValue = (sbyte) instruction.Operand;
+			else
+				return false;
+
+			return true;
+		}
+
+		static Instruction CreateLoadArraySizeOrOffsetInstruction (int intValue)
+		{
+			if (intValue < 0)
+				throw new ArgumentException ($"{nameof (intValue)} cannot be negative");
+
+			if (intValue < 9) {
+				switch (intValue) {
+				case 0:
+					return Instruction.Create (OpCodes.Ldc_I4_0);
+				case 1:
+					return Instruction.Create (OpCodes.Ldc_I4_1);
+				case 2:
+					return Instruction.Create (OpCodes.Ldc_I4_2);
+				case 3:
+					return Instruction.Create (OpCodes.Ldc_I4_3);
+				case 4:
+					return Instruction.Create (OpCodes.Ldc_I4_4);
+				case 5:
+					return Instruction.Create (OpCodes.Ldc_I4_5);
+				case 6:
+					return Instruction.Create (OpCodes.Ldc_I4_6);
+				case 7:
+					return Instruction.Create (OpCodes.Ldc_I4_7);
+				case 8:
+					return Instruction.Create (OpCodes.Ldc_I4_8);
+				}
+			}
+
+			if (intValue < 128)
+				return Instruction.Create (OpCodes.Ldc_I4_S, (sbyte)intValue);
+
+			return Instruction.Create (OpCodes.Ldc_I4, intValue);
+		}
+
+		bool UpdateMarshalRegisterMethod (MethodDefinition method, HashSet<string> markedMethods)
```

Comment:
Aside: this method gives me the heebie-jeebies and I have no idea if it's correct.

/me shudders.